### PR TITLE
fix(dsa): handle SM90 singleton score layouts and causal masks

### PR DESCRIPTION
**Pull request by jiayus-nvidia**
## Before submitting

- [x] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [x] I ran `pre-commit run` and committed any formatting changes.
- [x] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches and my changes comply.
- [x] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*`.

## Affected area

FE OSS kernels or CuTeDSL — SM90 DSA dense indexer and attention score recompute.

## Summary

Fix the existing SM90 singleton-layout and masked-output bugs exposed by the expanded varlen tests in #1147, including the matching attention-output inconsistency:

- Normalize the unused sequence stride of singleton `(B, H, 1)` weights/LSE through a metadata-only view, at both compilation and launch. PyTorch can consider this layout contiguous while retaining a final stride of `H`, which fails CuTe's unit-stride requirement.
- Write `-inf` for masked indexer and attention scores within visited KV tiles, matching the documented dense-score contract, SM100 behavior, and the prefilled values in skipped tiles. Attention masked positions still contribute zero to the L1 denominator.

Add indexer regression coverage for 16/32/64 heads, ratios 1/4, and singleton-first/last compilation orders; check scores, LSE, and reuse of one compile-cache entry. Add attention coverage for BSHD/THD, caller-provided output, singleton and fully masked rows, padding, skipped KV tiles, nonzero causal offsets, and 32/128 heads; compare all scores and L1 denominators.

## Why

[Pipeline 68834563, SM90 job](https://gitlab-master.nvidia.com/cudnn/cudnn_frontend/-/jobs/447989052) failed six `recompute=True` varlen cases: five stride errors and one mask mismatch. The affected implementation files were unchanged by #1147; its broader tests exposed the existing defects.

## Related issues

Related to #1147 — follow-up correctness fix for the SM90 failures exposed by its tests.

## API and compatibility impact

No signature or dependency changes. Singleton stride normalization changes only tensor metadata and adds no data copy or kernel launch. Masked output scores now consistently use `-inf` in both dense paths. Attention L1 reduction continues to exclude masked positions; its numerical contribution remains zero.

## Testing

H100 (SM90), CUTLASS DSL 4.6.2, CUDA 13.2, cuDNN 9.22.0. Tests imported the editable source checkout, verified in the pytest banner.

From `test/python`, with `PYTHONPATH` set to this checkout's `python` directory:

- `python -m pytest fe_api/dsa/test_DSA_indexer_forward.py fe_api/dsa/test_DSA_dense_score_recompute.py -k 'thd_varlen_tails or singleton_and_mask_cache' -n 0`: before the fix, **18 failed / 6 passed**; after the fix, **24 passed**.
- `python -m pytest fe_api/dsa/test_DSA_dense_score_recompute.py -k dense_attn_sm90_masked_output -n 0`: before the attention fix, **8 failed**; after it, **8 passed**.
- `python -m pytest fe_api/dsa/ -n 0`: **147 passed, 149 skipped, 134 deselected** after both fixes.
- `python -m pre_commit run --files python/cudnn/deepseek_sparse_attention/score_recompute/_interface_sm90.py python/cudnn/deepseek_sparse_attention/score_recompute/dense_score_recompute_sm90.py test/python/fe_api/dsa/test_DSA_dense_score_recompute.py`: passed.

## Summary by CodeRabbit

- **Bug Fixes**
  - Corrected dense attention and indexer outputs for causal-masked positions by returning negative infinity instead of zero.
  - Preserved zero denominator contributions for masked attention positions.
  - Improved single-query attention consistency across cached and uncached execution paths.
  - Preserved valid-position results while improving layout compatibility on supported hardware.

- **Tests**
  - Added regression coverage for batched and THD layouts, multiple head counts and ratios, masked outputs, padding masks, and in-place output reuse.

### CODE REVIEW

**Comment by jiayus-nvidia** [comment]
@cudnn-ci-bot run oss

**Comment by jiayus-nvidia** [comment]
@cudnn-ci-bot run oss

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `bd842e09358e` against base `8a6961c08ce9`.

The fixes make sense and I found no remaining actionable issue. Normalizing only the singleton sequence stride preserves addresses and applies on both compile-cache misses and hits. The final attention fix also resolves the earlier mask-consistency comment: visited and skipped masked columns now both follow the documented -inf contract, while denominator accumulation and the backward consumer still exclude those columns.

Validation: Independent exact-head H100 NVL run, DSL 4.7.0, CUDA 13.2 toolchain and cuDNN 9.25.1: DSA L0+L1 suite 151 passed, 254 skipped, 25 deselected. Verified the imported checkout and native-source identity before reusing the extension. Ran all 20 newly added regressions against the exact unfixed develop cb6b8e4 using the same environment: all 20 failed, with the expected singleton unit-stride errors and masked-score mismatches. All 20 pass in the current-head suite, covering both compile orders, multiple head counts, ratio masks, BSHD/THD, supplied output and denominators. H100 base/head/head/base spot check: BF16 B2/H32/KV256/D128, Q8 and Q128, both dense indexer and attention, with causal offsets making all columns valid. Analytic full-output/denominator checks and poisoned-output CUDA Graph replay passed in every arm. GPU medians changed by less than 0.4%; host enqueue remained about 37-38 us. No material regression observed in these four cases. Timings exclude compilation and separate host enqueue from device work. Current-head pre-commit passed. The author has started OSS pipeline 68870585; its mirror commit identifies the full bd842e09358e head, and it is still running. The older green pipeline 68867114 is not being counted as final-head CI. No duplicate run was triggered. The base advanced during review only through the unrelated BSA merge #1153; the DSA diff and reviewed contracts are unchanged.

Limitations: Local GPU validation used DSL 4.7.0; the author's 4.6.2 results were not independently rerun locally. Final-head OSS CI remains pending. Performance coverage is the stated four all-valid H100 workloads, not a full shape sweep or a speed comparison for singleton inputs that failed on the base.

Approved: no P0 or high-risk P1 found. Remaining findings stay with the owner; merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
Final-head OSS CI is now complete: pipeline 68870585 passed all 24 executed jobs, with 26 skipped and 6 manual jobs. Its mirror commit identifies the full reviewed `bd842e09358e4d0ece80442d0f060c58d04bee38` head. The SM90 OSS job passed 516 tests, including execution evidence from the expanded dense-score test file. This closes the pending-CI limitation in my review; no duplicate run was requested.

Codex bot review — model `gpt-6-astra`.
